### ISSUE TITLE
False positive arg-type when using typed lambda within a conditional

### ISSUE THREAD

**Issue by jamesbraza** [python/mypy#11711]
**Bug Report**

I have an `x: Optional[float]` passed into a function, and check `if x is not None`.  From this, mypy can understand that `x` is now a `float`.

However, when using the `x` inside a `lambda`, mypy seems to forget that `x: float`, it thinks `x: Optional[float`.

**To Reproduce**

Here is my minimal repro:

```python
from typing import Optional, Callable

def foo(threshold: Optional[float] = None) -> None:
    if threshold is not None:
        reveal_type(threshold)  # note: Revealed type is "builtins.float"

        _: Callable[[float], bool] = lambda num: num < abs(threshold)
        # error: Argument 1 to "abs" has incompatible type "Optional[float]";
        # expected "SupportsAbs[float]"  [arg-type]
```

**Expected Behavior**

I expect mypy not to raise an error.

**Actual Behavior**

```
7:60: error: Argument 1 to "abs" has incompatible type "Optional[float]"; expected "SupportsAbs[float]"  [arg-type]
```

**Known Workaround**

Utilizing `typing.cast` like so:

```python
from typing import Optional, Callable, cast

def foo(threshold: Optional[float] = None) -> None:
    if threshold is not None:
        reveal_type(threshold)  # note: Revealed type is "builtins.float"

        _: Callable[[float], bool] = lambda num: num < abs(cast(float, threshold))
```

No more errors are output here.

**Your Environment**

- Mypy version used: `0.910`
- Mypy command-line flags: nothing notable
- Mypy configuration options from `mypy.ini` (and other config files): nothing notable
- Python version used: `3.8.12`
- Operating system and version: macOS Big Sur version 11.5.1

**Comment by hauntsaninja** [comment]
See https://mypy.readthedocs.io/en/stable/common_issues.html#narrowing-and-inner-functions
Closing as duplicate of #2608
